### DISCUSSION TITLE
AIR Runtime 33.1 and AIR SDK 33.1 - is "stacktrace" output supported in production releases?

### DISCUSSION THREAD

**Discussion by TaS-Sensys** [airsdk/Adobe-Runtime-Support#2235] [General]
Hi,

I am able to get stacktrace output from the "adb.exe" launched applications with AIR and SDK 33.1, BUT only get
"null" as the stacktrace when running the "production release"

From my reading of older Adobe documentation, I should be able to get a stacktrace without line numbers, which would
be a big improvement over "null".

(https://stackoverflow.com/questions/4473059/how-can-i-get-stacktrace-for-adobe-air-global-runtime-errors-in-non-debug-mode)

I WANT the stacktrace so application users can report  "UncaughtErrorEvents()" popping up deep inside the code that are currently
silently corrupting the running software. 

(https://stackoverflow.com/questions/16073357/actionscript-compiler-uncaughterrorevent)

Am I missing any configurations? I am using "<verbose-stacktraces>true</verbose-stacktraces>" in frameworks/air-config.xml and "-compiler.verbose-stacktraces=true" arguments in my ant build.xml

Thanks
-Todd

PS: similar issues using very old air compiler v 2.0.2, but this might be too old, but is still running against Air runtime 33.1!

**Comment by 2jfw** [comment]
Quite vague and I cannot tell for sure but we did fail for everything we tried (using compiler args like debug mode, verbose stack traces, ...).
I remember a discussion form the Flex mailing list where this was discussed and the conclusion was that you explicitly need to run the swf with adb and have those (compiler) flags set when building (so the relase build needs to be run with adb)

**EDIT:** We were never able to see the _line numbers_ in the stacktraces (stacktraces do work well!)

**Reply by TaS-Sensys** [reply]
Yes, I can confirm the release build will return stacktraces with adb (debug=false), but I need to confirm if the release build can support the feature with the production Harman/AdobeAir.exe installation.

(I don't know if the feature ever worked pre Harman releases, from my reading it seems it did with Adobe Air 3.5+ or even earlier)

I am ALSO open to suggestions for stacktrace alternatives I could put in the UncaughtErrorEvent handler to indicate where in the code the problem occurred - this feature works well, I want my userbase to send me bugs/get prompted to restart the program to avoid corrupting behaviors that NOT catching these errors results in!

Thanks!

**Comment by Elintondm** [comment]
For us in our release build (debug=false) the stack traces work well but
the line numbers are not shown.

Anyone already tested if the debug version has a huge impact on performance?

Em ter., 25 de out. de 2022 às 14:10, TaS-Sensys ***@***.***>
escreveu:

**Reply by 2jfw** [reply]
Debug version has a very huge impact on performance - some random tests show at least half or less the performance of the release build.
Our huge (SeMSy 5) app (flex) e.g. runs with ~30fps in debug but will easily reach 60+fps in release.
Performance impact of course depends on the complexity and number of the things which are processed

**Comment by waveofthought-code** [comment]
fwiw i have been outputting stack traces via uncaught exception events for literally years -- first from flashplayer on the web, then from an Air captive runtime both in debug and release versions.   Running from adl you get line numbers, but packed up with adt you don't.   My release versions don't have any switch for "debugging = true" anywhere so i've always considered this an intrinsic feature (perhaps ignorantly) .

I publish my main swf via Animate, but the loader (and packaging) is built via VSCode.

**Comment by TaS-Sensys** [comment]
Thank you for the replies - just like the Internet, it seems the stacktraces are working in the release builds for some users but not others!

To be specific, my AIR targets are the Windows-based AIR Runtimes

**Comment by TaS-Sensys** [comment]
So to verify the flags I am using, I built with

(in build.xml)
\<arg value="-debug=false"/\>
and
\<arg line="-compiler.verbose-stacktraces=false" /\>

and my binary size went down to 
3884351 Nov  2 12:11 Main.swf
from
5500000+ bytes


WHEN I launch the app with Harman "adl", I can capture uncaught errors (here generated by mouseclick)
and get a simple stacktrace (no numbers) - I get the built-in
![2022-1102_simplestacktrace_harman-adb](https://user-images.githubusercontent.com/104396288/199595498-6363710f-df4e-4396-84c4-7b4fe627dd76.png)

and then from my own popup, I also post the stack trace:
![2022-1102_stacktrace-harman-adb](https://user-images.githubusercontent.com/104396288/199596108-3bad95d1-18dd-451b-8b72-ce6fe63deaa0.png)


WHEN I package the Main.swf into an .exe installable package to run under the Harman Runtime, the
popup alert still appears as in the "adl" run but the stacktrace content posts as NULL:

![2022-1102_alert_stacktrace_null](https://user-images.githubusercontent.com/104396288/199595749-b01d3552-f0af-481e-a160-6f7cbca29ba5.png)


(I assume there is no built-in/native error window by design?)


ARE there other compiler flags or entries in the air-config.xml or flex-config.xml that could be DISABLING the stacktraces
when I run under the production Harman-Adobe Air Runtime?

It seems to work for some of you so I remain hopeful!

Thank you
-Todd

**Reply by TaS-Sensys** [reply]
my uncaught capture code, and error generator!


		public function onClickMiddleError(e:MouseEvent):void {
			if (trafficdotSystem.superUserEnabled) {
				trace("MainMediator.as: onClickMiddleError(): MIDDLE, trigger Error()");
				throw new Error();
			} else {
				trace("MainMediator.as: onClickMiddleError(): MIDDLE, need superuser to trigger Error()");
			}
		}

		public function uncaughtErrorHandler(e:UncaughtErrorEvent):void
		{
			if (trafficdotSystem.superUserEnabled) {
				trace("MainMediator.as: uncaughtErrorHandler(): e=" + e.error.getStackTrace());
				TdAlert.show("Uncaught Error Detected:\n\nstack:\n\n" + e.error.getStackTrace() + "\n\nerror:\n\n" + e.toString(), null, null, true, true);
			} else {
				trace("MainMediator.as: uncaughtErrorHandler(): caught, popup w superuser");
			}
		}

**Reply by TaS-Sensys** [reply]
and the listener call:


			view.root.loaderInfo.uncaughtErrorEvents.addEventListener(UncaughtErrorEvent.UNCAUGHT_ERROR, uncaughtErrorHandler);

**Comment by TaS-Sensys** [comment]
Bueller? Bueller? Got a nice feature if I get this output going :)

**Comment by TaS-Sensys** [comment]
okay, answering my own question - the fix is a bit of a hack, but MAYBE SOMETHING THAT CAN BE ADDED TO THE RUNTIME?

THIS works for me UNDER WINDOWS 10
from
https://stackoverflow.com/questions/4473059/how-can-i-get-stacktrace-for-adobe-air-global-runtime-errors-in-non-debug-mode

1. Windows: Browse to its install directory in a file explorer. 
2. Navigate to {app-folder}▸META-INF▸AIR▸. Here you can find a file called hash. 
3. Duplicate the hash file and rename it to debug. 
4. Open the debug file with some text editor and remove the content.
5. Done now you get the stack trace + line numbers.


Again, how/could this be added to the runtime? I might modify my installer to automate this approach.

**Reply by ajwfrost** [reply]
Hi

There's two separate things here, build time and run time.

File and line number information is only available in a SWF file if it's generated with debug information in it. If you view the SWF in a decompiler you'll see all the extra information about which file was used (including full path on your machine) and a 'debugline' command every few AS3 instructions so that this stack trace can be generated. If you generate a SWF without debug information, then your stack traces cannot include file/line details.

Then at runtime, whether you get a stack trace or not depends on whether the runtime is tracking them. It does by default if you launch using ADL i.e. debugging from an IDE. It doesn't by default if you launch from a generated executable (i.e. when a .air file is installed, or when a bundle or other package is created and installed/run).

So I think what you're saying is that you'd like an application to be able to track the call stack even when it's a production-type app... and having a 'debug' file there is one way of forcing this? It may also be doing other things and adding to the overhead of the runtime, it basically means that on each method call there would need to be extra steps taken to work out the method name and push that onto the stack.. which implies its injecting extra calls in when JIT-compiling a method. We can check here though (since JIT compilation is also used when debugging with ADL so there's no reason why it can't, it's just whether you want this or not...)

If folk are already trying to hack this capability, it seems to make more sense to just add a flag to the app descriptor file to say that an application should always generate stack traces... :-)

[Edit] just to note though, given the above, even if there's a flag to get it to generate stack traces, there's nothing you can do to output file/line information there if the SWF didn't have those in when it was compiled... I would think that the 'verbose stack trace' argument means you include file/line details, but the rest of the 'debug' capabilities perhaps would be excluded... we can have a play... [/Edit]

thanks

**Reply by 2jfw** [reply]
I would imagine the following:
- (Release) app build _with_ `-debug` compiler argument to generate stack traces _with_ line numbers and file (e.g.
`at mx.collections::ListCollectionView/addItemsToView()[D:\flex_sdk\Sources\flex-sdk\frameworks\projects\framework\src\mx\collections\ListCollectionView.as:1168]`)
- (Release) app build _without_ `-debug` compiler argument to generate only stack traces (e.g.
`at mx.collections::ListCollectionView/addItemsToView()`)
- _No_ requirement to run the app with ADL to get the full stack traces with line numbers
- _Not_ using `App Descriptor` to determine if it is a debug build (/generate line numbers) - reason: it would simply require more effort to maintain two App Descriptor files, especially on CI/automated build servers
- `-verbose-stacktraces` compiler argument to be deprecated

**Reply by ajwfrost** [reply]
@2jfw this would mean that stack traces would be enabled for all normal release apps? I'm not sure most folk would want this increased overhead?

For the app descriptor change, I didn't mean that you'd need to maintain two different descriptors, I just meant that if you wanted to have the new capability of stack traces being available from a release package, then you'd need to add a request in the app descriptor. But that flag would only be used by the release apps, it wouldn't stop/affect the stack traces from working when you run from ADL; and stack trace information (when it's present, regardless of anything else) would include line/file information only if this is available within the SWF, so that's then the compile-time option.

**Reply by 2jfw** [reply]
Mm, I am now wondering about the fact that our release non-debug app the Uncaught Error handler does actually always have the full stack trace accessible..

**Reply by 2jfw** [reply]
Regarding the app descriptor I see what you mean - its just about accessing the new capability rather than actually instructing the compiler to build with it

**Reply by TaS-Sensys** [reply]
> Mm, I am now wondering about the fact that our release non-debug app the Uncaught Error handler does actually always have the full stack trace accessible..

EXACTLY! All the stacktrace data is in the application despite not compiling for it!

1) In my build.xml, 
compiler has
-debug=false
and
-verbose-stacktraces=false



2) I generate a <app>.exe installer program - this is in my build.xml, and I am trying to see if there are flags for a runtime debug here?

		<exec executable="C:\Program Files (x86)\Inno Setup 6\iscc.exe">
		 <arg value="TD2Setup.iss" />
		</exec>
(I will be digging into this program next - it builds our installer .exe)

3) Install the *.exe program by running the <app>.exe

4) I follow the slashdot recipe above - the simplified version is add an empty file "debug" to <air program name>/META-INF/AIR/debug

5) Run the installed AIR <app>

6) Generate an error/stacktrace (I trigger middle mouse button event for this) AND it works - no line numbers (that is expected with -verbose-stacktrace=false) but all the information required to generate a useful (to me, names are great) stack trace with names are present with -debug=false and -verbose-stacktrace=false (recall too that verbose-stacktrace=true app was ~2x bigger than =false)


![2022-1111_td2_empty-debug-file](https://user-images.githubusercontent.com/104396288/201399088-e1170313-e517-43f7-b87a-f85033bb46ac.png)


SO I am looking for a flag or process to enable these traces since my air program has them available anyway.
BUT yes, perhaps execution is impacted?

THIS all said, adding this empty "debug" file and restarting the application will provide my field engineers/advanced a process
to collect stack traces of uncaught errors in the meantime. I might add the debug file creation to the Air program itself when enabling a "stack collection" mode (will require a restart after the "debug" file is created)...

Thanks!

**Reply by 2jfw** [reply]
> @2jfw this would mean that stack traces would be enabled for all normal release apps? I'm not sure most folk would want this increased overhead?

@ajwfrost Could you comment on the callstacks being available in release build apps, though `-debug=false` and `verbose-stacktraces=false` is set? If I understand you correctly, this should not be the case (?)

**Reply by ajwfrost** [reply]
@2jfw yes I'm also wondering how you've ended up with that..! I didn't think it should be possible; it may depend a little on how you packaged up the application. Are you able to send it to us to look at? not sure what platform it is for, but there's a possibility perhaps that there is a 'debug' file there generated by the ADT tool which may be impacting it?

**Reply by 2jfw** [reply]
If you still have the app I provided from a couple of weeks ago you could check there - if not, it is no problem for me to re-upload it..

**Reply by TaS-Sensys** [reply]
@ajwfrost I can upload a Main.swf and then the packaged version of the Application - could you decompile/find the stack traceable code within if it exists? I built you a branch to so you can trigger a stack trace via middle mouse button. This version CREATES the "META-INF/AIR/debug file at runtime if the preference is set to catch uncaught errors! (so must run as admin if fully installed to create OR remove the "debug" file!!)

Older publicly released code version is here on our site https://go.sensysnetworks.com/trafficdot in a zip if that is enough...

**Reply by TaS-Sensys** [reply]
@ajwfrost and @2jfw, I created a github repository and placed the artifacts from my dev build there for your review as "collaborators" - ie stacktrace contents with -debug=false and -verbose-stacktraces=false (results in Main.swf) and then atd.bat packeged exe file in TrafficDOT2.exe.

BY default this version enables the "uncaughtErrorException()" listener that can be triggered by the middle mouse button at the root of the application. This process NOW also creates the "<app root>/META-AIR/AIR/debug" file, so running a full TrafficDOT2.exe install needs to run as administrator to add/remove the file from the "Windows Programs (x86)/" directory.

Any feedback appreciated!

**Reply by 2jfw** [reply]
@TaS-Sensys Thanks! However I cannot really assist with the version you provided but I can (re)provide @ajwfrost you a version of our release build with captive runtime where the stack traces on uncaught error handler are fully available (without line numbers)

**Reply by TaS-Sensys** [reply]
@2jfw was hoping you guys had some "de-compilers" available to confirm how much overhead may be within. Skipping
verbose-stacktraces removes 2 of 5 MB in the Main.swf! Even so, the smaller Main.swf returns trace info (file and function) though without line numbers.

**Reply by ajwfrost** [reply]
Yes we have some decompiling options ... and have just gone through a bunch of different combinations with different compilers.

I'm glad to say that we're seeing the same behaviour in AIR 32 as in AIR 50, which basically seems to ignore the "verbose stack traces" option. If you request a debug swf, you get file details and line numbering within the ActionScript byte code. If you request a non-debug swf, you don't.

The difference comes if you use the Flex compiler, where if you build a debug swf will include the file details and line numbered, but if you request a non-debug swf then there is a difference coming dependending on the use of verbose-stacktraces. If you have this false then you don't get the file/line info, but if you set this true then you do get the file/line info, but the swf is smaller than the debug versions and doesn't have the extra debugger instructions.

Will check the runtime behaviours with the three varieties of files...

**Reply by ajwfrost** [reply]
ADL with debug:
SWF has file/line and debugger info -> stack trace shows file/line numbers
SWF has file/line but no debugger -> stack trace shows file/line numbers
SWF has no file/line or debugger -> stack trace with functions but no file/line numbers

ADL with -nodebug:
This appears to only affect the Flex-built SWFs! I don't get any stack trace with the Flex-built SWF even if it's debugger enabled.
For an AIR-built SWF I'm always getting a stack trace but without file/line number info even if it's available in the SWF.

Captive runtime application:
Flex app -> no stack trace
AIR-built SWF -> stack trace with functions but no file/line numbers


Does this match what people are seeing? It looks like there are two differences if you're using Flex:
a) 'production' versions don't display any stack traces
b) there's an option to have stack traces without the debugger info .. but this only changes things when you're using ADL?!

But for an app built with AIR, even without debug info, it should be that you're seeing call stack information even in a deployed app.

The one thing I hadn't checked is for an application using the shared runtime, but I would expect that behaviour to be the same as for the captive runtime apps. I'm using Windows btw, in case you're seeing different results on different platforms..!

**Reply by ajwfrost** [reply]
FYI found a potential reason why sometimes there are stack traces in release builds .. this behaviour was changed based on the SWF version: if it's < version 18 then it doesn't create stack traces unless there's a debugger attached.

The functionality within the Error object to get the stack trace looks like it has different behaviour depending on whether you have a debugger attached: if you do, there will already be a stack trace and collection of call stack nodes, and the object is just quickly created. If there's no debugger then it will build up the stack by working through the method frames and creating the object.

See https://github.com/adobe/avmplus/blob/master/core/AvmCore.cpp#L4926 

So actually, if you have the debugger set-up, it would work for your apps that are using an older SWF version code, but is unnecessary for apps that are using a newer SWF version.

I just checked back and the "Flex" SWFs that I built using Flex SDK have a swf version of 14. So that would explain it then, nothing to do with Flex and everything to do with the version code and the compatibility level being using within the AIR runtime..


Does anything need to change here? The only thing I am wondering is whether you'd want a release build of the runtime to display the filenames/line numbers should the SWFs contain those details. That would need some changes in the runtime and JIT code but could be added if there's a big need for it...

**Reply by 2jfw** [reply]
I think the three following constellations would make sense then:
1) Release Build with no stack trace (and thus also no line numbers/file names) -> Default setting, no compiler args required
2) Release Build with stack trace but no line numbers/file names -> Compiler arg `-verbose-stacktraces=true`
2) Release Build with stack trace and line numbers/file names -> Compiler arg `-debug=true` (no need for `verbose-stacktraces`)

> The only thing I am wondering is whether you'd want a release build of the runtime to display the filenames/line numbers should the SWFs contain those details.

I find it confusing to differentiate between "release build of runtime" and "SWF" - for me, the SWF is part of the build process of a release build. Am I mistaking here something?

**Reply by ajwfrost** [reply]
I think the slight challenge with your first suggestion is that this would change the default behaviour back from what it is now, to something that it was several years ago - and obviously there was a reason that Adobe changed that, so I'm a little reluctant to revert it...

I'd also say that "verbose" stack traces would imply having file information and line numbers whereas "non-verbose" stack traces are just the functions, so the second one doesn't quite sit right to me either!

But yes, there is a difference here between the SWF build and the runtime. The SWF contains instructions for the runtime, so if the SWF says that it should enable debugging, then the runtime goes checking for a debugger.. and then the runtime behaviour is also different based on whether or not a debugger is actually found/connected. I think it's safe to say that for people who are actually debugging, they would (a) be building a SWF with debug mode (hence we'd have file/line info in there) and (b) they'd be using ADL with the expectation of it connecting to a debugger.

By the "release build" of a runtime I mean when you're not using the ADL tool but are using what would actually be run when you're using a deployed application, i.e. an application .exe file on Windows, or a .app bundle on Mac, etc. So there is a potential need to debug things in this scenario, just because there are differences between running via ADL and running via this (AppEntry) mechanism.

Going back to your three options, I agree with these as options, but the configuration settings to get to them I think need to be different:
1. Release Build with no stack traces: this is a change to existing behaviour so I would think we need a custom setting/option for this, i.e. "opt in to remove stack trace capability" hence preserving the default/existing behaviour.
2. Release Build with stack trace but no line numbers/file names -> default setting, as it is now, as long as you have SWF v18+
3. Release Build with stack trace and line numbers/file names -> we could either have this as an opt-in setting, or we could just add line numbers/file names in any call stack where the SWF contains these. And then, to get these, you would either use the "-debug" option, or (if we implemented it in the AIR compiler) the "-verbose-stacktraces" option.

My preference I think, looking at it, would be for applications to specify what behaviour they want via a configuration item e.g.
`<stacktraces>none|standard|verbose</stackstraces>`
This would be the simplest way of doing it and means that we can avoid changing things for existing applications if they're just updating onto a later runtime.

**Reply by 2jfw** [reply]
Sounds good :) Also thanks for the explanation regarding runtime and SWF, I do now understand what you are meaning.

So, where would the configuration item be placed?

**Reply by TaS-Sensys** [reply]
> Does anything need to change here? The only thing I am wondering is whether you'd want a release build of the runtime to display the filenames/line numbers should the SWFs contain those details. That would need some changes in the runtime and JIT code but could be added if there's a big need for it...

For my OLD Flex compiler 4.1x output then (and I will be staying with this ancient one - your suggestion ;) ), I am leveraging the presence of the empty "debug" file to trigger  "filename/function"  stacktraces in the current(ie Harman) AIR Runtime. I built a "toggle" to enable/disable the file, requiring the user to "Run As Administrator" and restart the application for the mode to take place. 

MY runtime "feature request" would be to have a non-file mechanism to enable or disable the "debug" state of the AdobeAIR.exe runtime. Not sure if there is a precedent for any other flags passed into the AdobeAIR.exe runtime?

A remaining question is how much overhead does the "debug" state of the runtime add to the code, if any? Is the dynamic tracking
of file/function occurring all the time or just under runtime "debug"? I can do some timing tests here too, we are debating leaving this 
"debug" file/state on as the default. Could I actually speed up my code using version 18+ of the SDK or would my Flex app not be affected? 

Thank you for the analysis!

**Reply by ajwfrost** [reply]
In your case, I think the debug file would trigger the full 'debugger' capabilities within the runtime, which means that you'd potentially get a little bit more overheads at each function call - but the worst bit might be at start-up, if the runtime is sat briefly to see if it can connect to an active debugger process (fdb).

So hopefully it would be possible to leave out this file but to get the stack traces by setting this new configuration flag to standard/verbose..?

I am thinking this would be in the XML configuration file as a top level option...

thanks

**Reply by ajwfrost** [reply]
FYI, added this in the latest release, https://airsdk.harman.com/download/50.1.1.1

**Comment by bobaoapae** [comment]
are this new config <stacktraces> working? i don't find anything related in the releases notes. @ajwfrost

**Reply by ajwfrost** [reply]
Hi
From the above link to the 50.1.1.1 download, you can get at the relevant release notes (search then for "2235" for this issue..)
Relevant text:

> A new app descriptor setting has been introduced at the top level, “<stacktraces>”, which can be set to one of the following values:
• none: no stack traces will be available in release builds
• standard: a stack trace listing the call stack will be provided
• verbose: a stack trace that includes file name and line number information (where available from the SWF) will be provided
If the item is missing, the runtime falls back to the previous default behaviour which is to use “none” for pre-v18 SWFs and “standard” for SWF v18+

And I just checked, we even have updated the documentation for the app descriptor elements:
https://airsdk.dev/docs/building/application-descriptor-files/elements/application#stacktraces

thanks

**Reply by bobaoapae** [reply]
Thanks

**Reply by Menard1965** [reply]
Thx a lot. I can see the line numbers now, but is there a trick or method to write the Error to a file for later debug?

So if a customer has problemes with an APP i could lead them to a logfile on their harddrive ?
At the moenmt I use uncoughterror Handler to get the Info. But in the Adobe runbtime errors are not shown ?

Thx a lot
Michael
